This PR implements the protected methods check_ith_link(ith_link) and  check_q_vec(q_vec)

juanjqo opened the pull request:
# Main instructions

By submitting this pull request, you automatically agree that you have read and accepted the following conditions:

- Anyone wanting to propose a new modification or introduce new functionality should reach out to the team first, as proposed modifications that do not comply with the library's development philosophy and style, do not follow the library's architecture, do not introduce a clear and general benefit to the library other than to the person who proposed the modification will likely be rejected with no further discussion.
- Support for DQ Robotics is given voluntarily and it is not the developers' role to educate and/or convince anyone of their vision.
- Any possible response and its timeliness will be based on the relevance, accuracy, and politeness of a request and the following discussion.
- You are familiar with the [development workflow](https://github.com/dqrobotics/matlab/blob/master/CONTRIBUTING.md).
- Each pull request should contain only individual changes (several changes of the same type **are allowed** on the same pull request).
- Refactoring or modifying internal implementation that is working is not allowed unless comprehensively discussed with and approved by @bvadorno and @mmmarinho.

## Description of changes
![Static Badge](https://img.shields.io/badge/PR_type%3A-Pending%20tasks%20for%20the%2023.04%20release-blue)

![PR](https://github.com/dqrobotics/matlab/assets/23158313/21ff4789-6d55-45ec-b8c7-7e9aa5b1007a)
@dqrobotics/developers 

Hi @bvadorno, this PR implements the protected methods `check_ith_link(ith_link)` and  `check_q_vec(q_vec)`.  Furthermore, the methods `DQ_SerialManipulator.fkm()`, `DQ_SerialManipulator.pose_jacobian()`, and `DQ_SerialManipulator.pose_jacobian_derivative()` are updated to test if both the index and the dimension of the vector joints are valid. The rationale behind this is twofold:

1. Test the new methods `check_ith_link(ith_link)` and  `check_q_vec(q_vec)`. 
2. Fix an unexpected behavior in the `DQ_SerialManipulator` class

Example (of the unexpected behavior):

Consider this minimal example with a 7-DOF serial manipulator, in which the user tries to compute the pose Jacobian derivative to the joint 9 (An invalid joint, therefore I would expect an exception). 

```matlab
clear all
close all
clc

robot = FrankaEmikaPandaRobot.kinematics();
q = zeros(7,1);
J = robot.pose_jacobian(q,9);  % The index 9 is no valid!
disp('OK')
```

Output (Current Master Branch):
```matlab
OK
>> 
```


Output (Proposed Modifications):
```shell
Error using DQ_Kinematics/check_ith_link
Tried to access link index 9 which is unnavailable.

Error in DQ_SerialManipulator/pose_jacobian (line 325)
                obj.check_ith_link(ith);
>> 
```


Best regards, 

Juancho

bvadorno reviewed (changes requested):
@juanjqo, `check_q_vec()` makes sense in `DQ_Kinematics.m`, but `check_ith_link()` should be an abstract method in this class and implemented in each of its subclasses. This is because from the documentation of several methods in `DQ_Kinematics.m`, 

> Optionally, in case of coupled kinematic chains, FKM(q, to_ith_link) 
        % calculates the forward kinematic model up to the ith element in
        % the chain.

Therefore, your implementation does not work for classes other than the ones inheriting from `DQ_SerialManipulator` because in coupled kinematic chains (e.g., mobile manipulator), `ith=1` means the mobile base, whereas `ith=2` means the manipulator. 

Alternatively, `check_ith_link()`  can be implemented directly in `DQ_SerialManipulator`, meaning it won't be available for classes at the same hierarchical level. This is an easy, albeit partial, solution, but it's better now because you need only to move `check_ith_link()` to `DQ_SerialManipulator`. We can later implement it for `DQ_SerialWholeBody`.

juanjqo commented:
Hi @bvadorno, thanks for your feedback. I implemented the partial solution you mentioned. This is, I moved the method `check_ith_link()` from `DQ_Kinematics()` to `DQ_SerialManipulator()`.
